_posts/2019-04-24-lecture-27.md

```diff
@@ -0,0 +1,434 @@
+
+And the algorithm is summarized below:
+
+1. Split documents into $$P$$ partitions, with their corresponding topic mixture $$\delta_{di}$$
+2. Copy the topic description $$B$$ to each worker
+3. Each worker runs inference locally and update their parameters
+4. All workers send back the update to the master server, which aggregates the update
+5. Master server sends back the updated parameters to all workers, repeating the process.
+
+The problem with this approach is two-fold:
+
+1. The overall model is no longer guaranteed to converge, since ergodicity (fully-connectedness) is broken
+2. **Bigger problem**: communication time to send parameter updates between master server and the workers could be longer than the computation time.
+
+### CGS Model-parallel 1: GraphLab <d-cite key="Low2010"></d-cite><d-cite key="Gonzalez2012"></d-cite>
+
+To solve the communication problem, instead of copying the full model to all workers, we can instead have each worker access only (non-overlapping) portions of the model $$\rightarrow$$ model parallel method.
+The question is then: how do we split the model?
+
+Based on the GraphLab framework, we can represent the observations as a bipartite graph between documents and words. And then using vertex cut algorithm we can find partitions of the full graph, which can then be send to different workers to do inference locally in parallel. This is illustrated in the following figure:
+
+<img src="{{ '/assets/img/notes/lecture-27/graphlab.png' | relative_url }}" height="300px" />
+
+This helps eliminate the need to do global synchronization of $$B$$, since each worker now process disjoint sets of words (and documents). However, the vertex cut problem for graph partitioning is an NP-complete, and existing algorithms take significant amount of time to get good partitions. In this light, LightLDA was proposed.
+
+### CGS Model-parallel 2: LightLDA <d-cite key="Yuan2015"></d-cite>
+
+The idea of LightLDA is still to do model-parallel instead of data-parallel method, but using a much simpler partitioning compared to GraphLab framework. The idea is to split the matrix block-wise, and then rotate the blocks in each iterations, as illustrated below:
+
+<img src="{{ '/assets/img/notes/lecture-27/lightlda.png' | relative_url }}" />
+
+The caveat of this approach is on load imbalance: some documents could have a higher frequency of certain words, causing a worker to be overloaded compared to other workers. In a practical system, these things need to be addressed, for example, by grouping documents by word frequency.
+
+In summary, model-parallel approach is quite desirable over data-parallel for this task because it reduces dependency between workers, since the workers will not step on each other's toes while making updates locally, since the parameter space assigned to them are disjoint, unlike in data-parallel case.
+
+## Distributed Systems for ML
```

Comment:
Please format this section in paragraph form so that it reads more like a blog post, rather than bullet points from lecture slides. Here's an example that I liked: https://sailinglab.github.io/pgm-spring-2019/notes/lecture-13/

Reply:
@lileee Is this comment on line 387 on "Distributed Systems for ML" or the previous section? I feel the previous section is already in paragraph format.